src/mame/casio/ra3.cpp

```diff
@@ -13,67 +13,102 @@
 
 // device type definition
 DEFINE_DEVICE_TYPE(CASIO_RA3, casio_ra3_device, "casio_ra3", "Casio RA-3 RAM cartridge")
+DEFINE_DEVICE_TYPE(CASIO_RA6, casio_ra6_device, "casio_ra6", "Casio RA-6 RAM cartridge")
 
 /**************************************************************************/
-casio_ra3_device::casio_ra3_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
-	: device_t(mconfig, CASIO_RA3, tag, owner, clock)
+casio_ram_cart_device::casio_ram_cart_device(const machine_config &mconfig, device_type type, unsigned max_size, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, type, tag, owner, clock)
 	, device_memcard_image_interface(mconfig, *this)
+{
+	m_max_size = max_size;
+	m_mask = max_size - 1;
+	m_size = 0;
+}
+
+/**************************************************************************/
+casio_ra3_device::casio_ra3_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: casio_ram_cart_device(mconfig, CASIO_RA3, 0x1000, tag, owner, clock)
 {
 }
 
 /**************************************************************************/
-void casio_ra3_device::device_start()
+casio_ra6_device::casio_ra6_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: casio_ram_cart_device(mconfig, CASIO_RA6, 0x4000, tag, owner, clock)
 {
-	m_ram.resize(0x1000, 0xff);
+}
+
+/**************************************************************************/
+void casio_ram_cart_device::device_start()
+{
+	m_ram.resize(m_max_size, 0xff);
 
 	save_item(NAME(m_ram));
+	save_item(NAME(m_size));
+	save_item(NAME(m_mask));
 }
 
 /**************************************************************************/
-std::pair<std::error_condition, std::string> casio_ra3_device::call_load()
+std::pair<std::error_condition, std::string> casio_ram_cart_device::call_load()
 {
-	const size_t size = m_ram.size();
-	if (length() != size)
+	const u32 size = loaded_through_softlist() ? get_software_region_length("rom") : length();
+
+	// size must be a power of two and at least 4kb
+	if (size < 0x1000 || (size & (size - 1)))
 		return std::make_pair(image_error::INVALIDLENGTH, std::string());
 
-	fseek(0, SEEK_SET);
-	const size_t ret = fread(m_ram.data(), size);
-	if (ret != size)
-		return std::make_pair(std::errc::io_error, "Error reading file");
+	// allow loading larger than the maximum size (e.g. for CZ-1 ROM carts that use oversized ROMs)
+	m_size = std::min(size, m_max_size);
+	m_mask = m_size - 1;
```

Comment:
Should this allow loading oversized RAM cartridges, or only ROM cartridges from the software list?

Reply:
Using oversized RAM carts works on real CZs too, since they all use the same connector.